File: CGXHorizontalCollectionView-OC/CGXHorizontalCollectionView/CGXHorizontalCollectionView.m
//
//  CGXCollectionViewGeneralView.m
//  CGXHorizontalCollectionView-OC
//
//  Created by CGX on 2019/10/01.
//  Copyright © 2019年 CGX. All rights reserved.
//

#import "CGXHorizontalCollectionView.h"
#import "CGXHorizontalCollectionLayout.h"
@interface CGXHorizontalCollectionView()

@property (nonatomic , strong , readwrite) NSMutableArray<CGXHorizontalCollectionModel *> *dataArray;
@property (nonatomic , strong , readwrite) UICollectionView *collectionView;

@property (nonatomic , strong )CGXHorizontalCollectionLayout *flowLayout;
@end

@implementation CGXHorizontalCollectionView

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self initializeViews];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self initializeViews];
    }
    return self;
}
- (void)willMoveToSuperview:(UIView *)newSuperview {
    [super willMoveToSuperview:newSuperview];
    UIResponder *next = newSuperview;
    while (next != nil) {
        if ([next isKindOfClass:[UIViewController class]]) {
            ((UIViewController *)next).automaticallyAdjustsScrollViewInsets = NO;
            break;
        }
        next = next.nextResponder;
    }
}
- (void)initializeViews
{
    self.backgroundColor = [UIColor whiteColor];
    self.row = 3;
    self.minimumLineSpacing = 10;
    self.minimumInteritemSpacing = 10;
    self.insets = UIEdgeInsetsMake(10, 10, 10, 10);
    self.showsHorizontalScrollIndicator = NO;
    self.space = 0;
    self.row = 3;
    self.section= 1;
    self.bounces = YES;
    self.stop = NO;
    CGXHorizontalCollectionLayout*flowLayout= [[CGXHorizontalCollectionLayout alloc] init];
    flowLayout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    flowLayout.stop = self.stop;
    self.flowLayout = flowLayout;
    self.collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(self.frame), CGRectGetHeight(self.frame)) collectionViewLayout:flowLayout];
    self.collectionView.delegate = self;
    self.collectionView.dataSource = self;
    self.collectionView.showsVerticalScrollIndicator = NO;
    self.collectionView.showsHorizontalScrollIndicator = self.showsHorizontalScrollIndicator;
    self.collectionView.bounces = self.bounces;
    self.collectionView.backgroundColor = self.backgroundColor;
    [self.collectionView registerClass:[CGXHorizontalCollectionCell class] forCellWithReuseIdentifier:NSStringFromClass([CGXHorizontalCollectionCell class])];
    [self.collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:NSStringFromClass([UICollectionViewCell class])];
    //给collectionView注册头分区的Id
    [self.collectionView registerClass:[UICollectionReusableView class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:NSStringFromClass([UICollectionReusableView class])];
    //给collection注册脚分区的id
    [self.collectionView registerClass:[UICollectionReusableView class] forSupplementaryViewOfKind:UICollectionElementKindSectionFooter withReuseIdentifier:NSStringFromClass([UICollectionReusableView class])];
    [self addSubview:_collectionView];
    if (@available(iOS 11.0, *)) {
        self.collectionView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentNever;
    }
    if (@available(iOS 10.0, *)) {
        self.collectionView.prefetchingEnabled = NO;
    }
    self.collectionView.alwaysBounceHorizontal = YES;
    self.collectionView.decelerationRate = UIScrollViewDecelerationRateFast;
}
- (void)layoutSubviews
{
    [super layoutSubviews];
    self.collectionView.backgroundColor = self.backgroundColor;
    self.collectionView.frame = CGRectMake(0, 0, CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds));
    [self.collectionView reloadData];
}
- (void)registerCellClass:(Class)classCell IsXib:(BOOL)isXib
{
    NSAssert(![classCell isKindOfClass:[UICollectionViewCell class]] , @"cell类型需要是UICollectionViewCell");
    if (isXib) {
        [self.collectionView registerNib:[UINib nibWithNibName:[NSString stringWithFormat:@"%@", classCell] bundle:nil] forCellWithReuseIdentifier:[NSString stringWithFormat:@"%@", classCell]];
        
    } else{
        [self.collectionView registerClass:classCell forCellWithReuseIdentifier:[NSString stringWithFormat:@"%@", classCell]];
    }
}
- (void)registerFooter:(Class)footer IsXib:(BOOL)isXib
{
    if (![footer isKindOfClass:[UICollectionReusableView class]]) {
        NSAssert(![footer isKindOfClass:[UICollectionReusableView class]], @"注册cell的registerCellAry数组必须是UICollectionReusableView类型");
    }
    if (isXib) {
        [self.collectionView registerNib:[UINib nibWithNibName:[NSString stringWithFormat:@"%@", footer] bundle:nil] forSupplementaryViewOfKind:UICollectionElementKindSectionFooter withReuseIdentifier:[NSString stringWithFormat:@"%@", footer]];
    } else{
        [self.collectionView registerClass:footer forSupplementaryViewOfKind:UICollectionElementKindSectionFooter withReuseIdentifier:[NSString stringWithFormat:@"%@", footer]];
    }
}
- (void)registerHeader:(Class)header IsXib:(BOOL)isXib
{
    if (![header isKindOfClass:[UICollectionReusableView class]]) {
        NSAssert(![header isKindOfClass:[UICollectionReusableView class]], @"注册cell的registerCellAry数组必须是UICollectionReusableView类型");
    }
    if (isXib) {
        [self.collectionView registerNib:[UINib nibWithNibName:[NSString stringWithFormat:@"%@", header] bundle:nil] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:[NSString stringWithFormat:@"%@", header]];
    } else{
        [self.collectionView registerClass:header forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:[NSString stringWithFormat:@"%@", header]];
    }
}

- (void)setBounces:(BOOL)bounces
{
    _bounces = bounces;
    self.collectionView.bounces = bounces;
}
- (void)setShowsHorizontalScrollIndicator:(BOOL)showsHorizontalScrollIndicator
{
    _showsHorizontalScrollIndicator = showsHorizontalScrollIndicator;
    self.collectionView.showsHorizontalScrollIndicator = showsHorizontalScrollIndicator;
}
- (void)setStop:(BOOL)stop
{
    _stop = stop;
    self.flowLayout.stop =stop;
    self.collectionView.collectionViewLayout = self.flowLayout;
    [self.collectionView.collectionViewLayout invalidateLayout];
    [self.collectionView reloadData];
}
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return 1;
}
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return self.dataArray.count;
}
- (NSMutableArray<CGXHorizontalCollectionModel *> *)dataArray
{
    if (!_dataArray) {
        _dataArray = [NSMutableArray array];
    }
    return _dataArray;
}
- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section
{
    return self.insets;
}
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section
{
    return self.minimumLineSpacing;
}
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section
{
    return self.minimumInteritemSpacing;
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    UIEdgeInsets insets =self.insets;
    NSAssert(self.row > 0, @"每行至少一个");
    NSAssert(self.section > 0, @"至少一行");
    
    CGFloat spaceV = (insets.top+insets.bottom);
    CGFloat spaceH = (insets.left+insets.right);
    
    CGFloat height = ((CGRectGetHeight(collectionView.frame)- spaceV) - (self.section - 1) * self.minimumLineSpacing) / self.section;
    
    CGFloat width = ((CGRectGetWidth(collectionView.frame)-spaceH) -self.space - (self.row -1) *self.minimumInteritemSpacing) / self.row;
    return CGSizeMake(width,floor(height));
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForFooterInSection:(NSInteger)section
{
        return CGSizeMake(50, CGRectGetWidth(collectionView.frame));
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section
{
        return CGSizeMake(30, CGRectGetWidth(collectionView.frame));
}
- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
{
    if (kind == UICollectionElementKindSectionHeader) {
        UICollectionReusableView *headview = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:NSStringFromClass([UICollectionReusableView class]) forIndexPath:indexPath];
        headview.backgroundColor = collectionView.backgroundColor;
        headview.backgroundColor = [UIColor orangeColor];
        return headview;
    } else {
        UICollectionReusableView *footview = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:NSStringFromClass([UICollectionReusableView class]) forIndexPath:indexPath];
        footview.backgroundColor = collectionView.backgroundColor;;
        footview.backgroundColor = [UIColor yellowColor];
        return footview;
    }
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    if (self.cellForItemBlock) {
        return self.cellForItemBlock(self, indexPath);
    }
    if (self.delegate && [self.delegate respondsToSelector:@selector(collectionHorizontalView:cellForItemAtIndexPath:)]) {
        return [self.delegate collectionHorizontalView:self cellForItemAtIndexPath:indexPath];
    }
    CGXHorizontalCollectionCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:NSStringFromClass([CGXHorizontalCollectionCell class]) forIndexPath:indexPath];
    CGXHorizontalCollectionModel *model = self.dataArray[indexPath.row];
    [cell updateWithModel:model];
    return cell;
}
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    if (self.didSelectItemBlock) {
        self.didSelectItemBlock(self, indexPath);
    }
    if (self.delegate && [self.delegate respondsToSelector:@selector(collectionHorizontalView:DidSelectItemAtIndexPath:)]) {
        [self.delegate collectionHorizontalView:self DidSelectItemAtIndexPath:indexPath];
    }
}
- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    if (self.scrollerBlock) {
        self.scrollerBlock(self, scrollView);
    }
    if (self.delegate && [self.delegate respondsToSelector:@selector(collectionHorizontalView:scrollViewDidScroll:)]) {
        [self.delegate collectionHorizontalView:self scrollViewDidScroll:scrollView];
    }
}
- (void)updateWihDataArray:(NSMutableArray <CGXHorizontalCollectionModel *> *)dataArray
{
    [self.dataArray removeAllObjects];
    for (CGXHorizontalCollectionModel *model in dataArray) {
        [self.dataArray addObject:model];
    }
    [self.collectionView reloadData];
}
- (void)updateWithModel:(CGXHorizontalCollectionModel *)model AtIndex:(NSInteger)index
{
    if (self.dataArray.count == 0) {
        return;
    }
    if (index>=self.dataArray.count) {
        return;
    }
    __weak typeof(self) weakSelf = self;
    [UIView animateWithDuration:0 animations:^{
      [weakSelf.collectionView performBatchUpdates:^{
          [weakSelf.dataArray replaceObjectAtIndex:index withObject:model];
          [weakSelf.collectionView reloadItemsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]]];
      } completion:nil];
    }];
    
}
/*
 // Only override drawRect: if you perform custom drawing.
 // An empty implementation adversely affects performance during animation.
 - (void)drawRect:(CGRect)rect {
 // Drawing code
 }
 */

@end

File: CGXHorizontalCollectionView-Swift/CGXCollectionHorizontalView/CGXHorizontalCollectionView.swift
//
//  CGXHorizontalCollectionView.swift
//  CGXHorizontalCollectionView-Swift
//
//  Created by CGX on 2019/10/01.
//  Copyright © 2019年 CGX. All rights reserved.
//

import UIKit

typealias ShowCellBlock = (_ baseView: CGXHorizontalCollectionView ,_ collectionView: UICollectionView,_ indexPath: IndexPath) -> UICollectionViewCell

typealias SelectCellBlock = (_ baseView: CGXHorizontalCollectionView,_ model:CGXHorizontalCollectionModel , _ indexPath: IndexPath) -> Void

typealias ScrollViewDidScrollBlock = (_ baseView: CGXHorizontalCollectionView,_ scrollView: UIScrollView) -> Void

typealias ShowHeaderBlock = (_ baseView: CGXHorizontalCollectionView ,_ headerViewAt:CGXHorizontalCollectionHeaderView,_ indexPath: IndexPath) -> UIView

typealias ShowFooterBlock = (_ baseView: CGXHorizontalCollectionView ,_ footerViewAt:CGXHorizontalCollectionFooterView,_ indexPath: IndexPath) -> UIView

class CGXHorizontalCollectionView: UIView,UICollectionViewDelegate,UICollectionViewDataSource,UICollectionViewDelegateFlowLayout {
    
    
    var showsVerticalScrollIndicator:Bool = false/*暂时水平 默认NO */
    var showsHorizontalScrollIndicator:Bool = false/*暂时垂直 默认NO */
    var bounces:Bool = false/*边距弹性 */
    var minimumLineSpacing: CGFloat = 10/*默认是10 */
    var minimumInteritemSpacing: CGFloat = 10/*默认是10 */
    var insets: UIEdgeInsets = UIEdgeInsets.init(top: 10, left: 10, bottom: 10, right: 10)/*默认是10,10,10,10 */
    var overFlowSpace: CGFloat = 20/* 一行溢出的item边距 */
    var row: Int = 3/*默认每行三个 */
    var section: Int = 1//默认一行
    
    var headerHeight: CGFloat = 0/*默认是0 */
    var footerHeight: CGFloat = 0/*默认是0 */
    lazy var headerColor:UIColor = { // 头分区颜色
        let color = UIColor.init(white: 0.93, alpha: 1)
        return color
    }()
    lazy var footerColor:UIColor = {// 脚分区颜色
        let color = UIColor.init(white: 0.93, alpha: 1)
        return color
    }()
    
    //实现代理
    open weak var delegate:CGXHorizontalCollectionViewDelegate?
    //点击的block
    fileprivate var selectCellBlock: SelectCellBlock?
    //展示的block
    fileprivate var cellForItemAtBlock: ShowCellBlock?
    //滚动block
    fileprivate var scrollViewBlock: ScrollViewDidScrollBlock?
    // 滚动方向
    fileprivate var scrollDirection:Bool = false
    //数据源。格式 @[xx,xx,xx,xx,xx,xx,xx]
    //数据源。 外界可读性 不可写入
    private(set)  lazy var dataArray: [CGXHorizontalCollectionModel] = [CGXHorizontalCollectionModel]()
    
    fileprivate func preferredFlowLayout() -> UICollectionViewLayout {
        let flowLayout = UICollectionViewFlowLayout()
        if (scrollDirection) {
            flowLayout.scrollDirection = .vertical;
        }else{
            flowLayout.scrollDirection = .horizontal;
        }
        return flowLayout
    }
    lazy var collectionView: UICollectionView = {
        let flowLayout = preferredFlowLayout();
        let  view = UICollectionView(frame: CGRect.init(x: 0, y: 0, width: frame.size.width, height: frame.size.height), collectionViewLayout: flowLayout)
        view.backgroundColor = self.backgroundColor;
        view.delegate = self;
        view.dataSource = self;
        view.showsVerticalScrollIndicator = false
        view.showsHorizontalScrollIndicator = false
        view.register(UICollectionViewCell.classForCoder(), forCellWithReuseIdentifier: "UICollectionViewCell")
        view.register(CGXHorizontalCollectionCell.classForCoder(), forCellWithReuseIdentifier: "CGXHorizontalCollectionCell")
        view.register(CGXHorizontalCollectionHeaderView.classForCoder(), forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "CGXHorizontalCollectionHeaderView");
        view.register(CGXHorizontalCollectionFooterView.classForCoder(), forSupplementaryViewOfKind: UICollectionView.elementKindSectionFooter, withReuseIdentifier: "CGXHorizontalCollectionFooterView");
        
        self.addSubview(view);
        return view
    }()
    open  func registerCell(_ classCell: AnyClass, isXib: Bool) {
        if !(classCell.self is UICollectionViewCell.Type) {
            assert(!(classCell.self is UICollectionViewCell.Type), "注册cell的必须是UICollectionViewCell类型")
        }
        if isXib {
            collectionView.register(UINib(nibName: "\(String(describing: classCell))", bundle: nil), forCellWithReuseIdentifier: "\(String(describing: classCell))")
        } else {
            collectionView.register(classCell, forCellWithReuseIdentifier: "\(String(describing: classCell))")
        }
    }
    override init(frame: CGRect) {
        super.init(frame: frame)
        dataArray = [CGXHorizontalCollectionModel]()
        self.delegate = self as CGXHorizontalCollectionViewDelegate
        self.collectionView.frame = CGRect.init(x: 0, y: 0, width: frame.size.width, height: frame.size.height)
        self.collectionView.reloadData()
    }
    
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
//MARK: - 数据源更新 数据初始化
extension CGXHorizontalCollectionView {
    //展示cell
    func cellForItemAtBlock(cellBlock:@escaping ShowCellBlock){
        cellForItemAtBlock = cellBlock
    }
    //点击cell
    func selectCellBlock(selectBlock:@escaping SelectCellBlock){
        selectCellBlock = selectBlock
    }
    //滚动
    func scrollViewBlock(scrollBlock:@escaping ScrollViewDidScrollBlock){
        scrollViewBlock = scrollBlock
    }
    // 横竖排列
    func scrollDirection(isVertical:Bool) -> Void {
        scrollDirection = isVertical
        self.collectionView.collectionViewLayout = preferredFlowLayout()
        self.collectionView.reloadData()
    }
    //更新数据
    func updateWithArray(dataArray:Array<CGXHorizontalCollectionModel>) -> Void {
        self.dataArray.removeAll()
        for model in dataArray {
            self.dataArray.append(model)
        }
        self.collectionView.reloadData()
    }
    
}
extension CGXHorizontalCollectionView {
    //分区数
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }
    //每个分区含有的 item 个数
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.dataArray.count;
    }
    //每个分区的内边距
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return insets
    }
    //最小 item 间距
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return CGFloat(minimumInteritemSpacing);
    }
    //最小行间距
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return CGFloat(minimumLineSpacing);
    }
    //item 的尺寸
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        let wSpace:CGFloat = insets.left + insets.right
        let hSpace:CGFloat = insets.top + insets.bottom
        let KWinW:CGFloat = collectionView.bounds.size.width
        let KWinH:CGFloat = collectionView.bounds.size.height
        
        if scrollDirection {
            //cell width
            let item_W:CGFloat    =  (KWinW - wSpace - CGFloat(row-1) * minimumInteritemSpacing) / CGFloat(row)
            //cell height
            let item_H:CGFloat    =   (KWinH - hSpace - CGFloat(section-1) * minimumLineSpacing-overFlowSpace)/CGFloat(section)
            return CGSize.init(width: floor(item_W), height: floor(item_H))
        } else {
            //cell width
            let item_W:CGFloat    =  (KWinW - wSpace - CGFloat(row-1) * minimumInteritemSpacing-overFlowSpace) / CGFloat(row)
            //cell height
            let item_H:CGFloat    =   (KWinH - hSpace - CGFloat(section-1) * minimumLineSpacing)/CGFloat(section)
            return CGSize.init(width: floor(item_W), height: floor(item_H))
        }
    }
    //每个分区区头尺寸
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
        if scrollDirection {
            return  CGSize.init(width: 0, height: headerHeight)
        }
        return  CGSize.init(width: 0, height: 0)
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForFooterInSection section: Int) -> CGSize {
        if scrollDirection {
            return  CGSize.init(width: 0, height: footerHeight)
        }
        return  CGSize.init(width: 0, height: 0)
    }
    //返回区头、区尾实例
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        if kind == UICollectionView.elementKindSectionHeader {
            let view:CGXHorizontalCollectionHeaderView = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: "CGXHorizontalCollectionHeaderView", for: indexPath as IndexPath) as! CGXHorizontalCollectionHeaderView
            view.backgroundColor = footerColor
            for v in view.subviews {
                v.removeFromSuperview()
            }
            let headerView = (self.delegate?.collectionHorizontalView(baseView: self, HeaderView: view, At: indexPath))!
            headerView.frame = CGRect.init(x: 0, y: 0, width: view.frame.size.width, height: view.frame.size.height)
            headerView.backgroundColor = headerColor
            view.addSubview(headerView)
            return view;
        } else {
            let view:CGXHorizontalCollectionFooterView = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: "CGXHorizontalCollectionFooterView", for: indexPath as IndexPath) as! CGXHorizontalCollectionFooterView
            view.backgroundColor = footerColor
            for v in view.subviews {
                v.removeFromSuperview()
            }
            let footerView = (self.delegate?.collectionHorizontalView(baseView: self, footerView: view, At: indexPath))!
            footerView.frame = CGRect.init(x: 0, y: 0, width: view.frame.size.width, height: view.frame.size.height)
            footerView.backgroundColor = footerColor
            view.addSubview(footerView)
            return view;
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        if cellForItemAtBlock.self != nil {
            return cellForItemAtBlock!(self,collectionView,indexPath)
        }
        return (self.delegate?.collectionHorizontalView(baseView: self, collectionView, cellForItemAt: indexPath))!

    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let model:CGXHorizontalCollectionModel = self.dataArray[indexPath.row]
        if self.selectCellBlock != nil {
            self.selectCellBlock!(self,model,indexPath)
        }
        self.delegate?.collectionHorizontalView(baseView: self, model: model, didSelectRowAt: indexPath)
    }
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        if self.scrollViewBlock != nil {
            self.scrollViewBlock!(self,scrollView)
        }
        self.delegate?.collectionHorizontalView(baseView: self, scrollView: scrollView)
    }
}

protocol CGXHorizontalCollectionViewDelegate: AnyObject {
    /**代理函数 cell的显示 */
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView,_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView, HeaderView:CGXHorizontalCollectionHeaderView,At indexPath: IndexPath) -> UIView
    
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView, footerView:CGXHorizontalCollectionFooterView,At indexPath: IndexPath) -> UIView
    /*点击事件*/
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView,model:CGXHorizontalCollectionModel, didSelectRowAt indexPath: IndexPath) -> Void
    
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView,scrollView: UIScrollView) -> Void
}
/// 提供JXSegmentedViewDelegate的默认实现，这样对于遵从JXSegmentedViewDelegate的类来说，所有代理方法都是可选实现的。
extension CGXHorizontalCollectionView:CGXHorizontalCollectionViewDelegate {
    
   // *代理函数 cell的显示
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView,_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell
    {
        let cell:CGXHorizontalCollectionCell = collectionView.dequeueReusableCell(withReuseIdentifier: "CGXHorizontalCollectionCell", for: indexPath) as! CGXHorizontalCollectionCell
        cell.contentView.backgroundColor = UIColor.white
        return cell;
    }
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView, HeaderView:CGXHorizontalCollectionHeaderView,At indexPath: IndexPath) -> UIView
    {
        return UIView.init()
    }
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView, footerView:CGXHorizontalCollectionFooterView,At indexPath: IndexPath) -> UIView
    {
        return UIView.init()
    }
    /*点击事件*/
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView,model:CGXHorizontalCollectionModel, didSelectRowAt indexPath: IndexPath) -> Void
    {
        
    }
    func collectionHorizontalView(baseView:CGXHorizontalCollectionView,scrollView: UIScrollView) -> Void
    {
        
    }
}
